### PR TITLE
TMDM-15182 Improper Restriction of XML External Entity Reference

### PR DESCRIPTION
**Pull request by hw-zhu**
https://jira.talendforge.org/browse/TMDM-15182
**What is the current behavior?** (You should also link to an open issue here)
Improper Restriction of XML External Entity Reference in kind of project

**What is the new behavior?**
Improper Restriction of XML External Entity Reference in kind of project

**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?

**What kind of change does this PR introduce?**

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

**Review by htyin — APPROVED** [review]

File: org.talend.mdm.commons.core/src/com/amalto/commons/core/utils/XMLUtils.java [hunk 1]
```diff
@@ -49,177 +45,31 @@
...
-     * @throws TransformerException
-     */
-    public static String getFirstTextNodeNotNull(Node contextNode, String xPath) throws TransformerException {
-    	String val=  getFirstTextNode(contextNode,xPath,contextNode);
-    	return val == null ? "" : val;
+            saxonTransformerFactory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+            saxonTransformerFactory.setAttribute("http://saxon.sf.net/feature/version-warning", Boolean.FALSE);
+            saxonTransformerFactory.setAttribute("http://saxon.sf.net/feature/allow-external-functions", Boolean.FALSE);
+            saxonTransformerFactory.setAttribute("http://saxon.sf.net/feature/trace-external-functions", Boolean.FALSE);
+//            saxonTransformerFactory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
```
Comment (zbcui) [inline org.talend.mdm.commons.core/src/com/amalto/commons/core/utils/XMLUtils.java:66]: Remove the commented code

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/DefaultStorageClassLoader.java [hunk 2]
```diff
@@ -90,7 +90,11 @@ public InputStream generateHibernateMapping() {
             DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
             factory.setNamespaceAware(true);
             factory.setExpandEntityReferences(false);
+            factory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, Boolean.TRUE);
+            factory.setFeature("http://xml.org/sax/features/external-general-entities", false);
+            factory.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
             DocumentBuilder documentBuilder = factory.newDocumentBuilder();
+//            DocumentBuilder documentBuilder = MDMXMLUtils.getDocumentBuilder().get();
```
Comment (zbcui) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/DefaultStorageClassLoader.java:97]: could remove

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/DefaultStorageClassLoader.java [hunk 3]
```diff
@@ -170,8 +174,11 @@ public Document generateHibernateConfiguration(RDBMSDataSource rdbmsDataSource)
         DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
         factory.setNamespaceAware(true);
         factory.setExpandEntityReferences(false);
-
+        factory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, Boolean.TRUE);
+        factory.setFeature("http://xml.org/sax/features/external-general-entities", false);
+        factory.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
         DocumentBuilder documentBuilder = factory.newDocumentBuilder();
+//        DocumentBuilder documentBuilder = MDMXMLUtils.getDocumentBuilder().get();
```
Comment (zbcui) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/DefaultStorageClassLoader.java:181]: could remove

File: org.talend.mdm.core/src/com/amalto/core/storage/SystemStorageWrapper.java [hunk 4]
```diff
@@ -339,7 +340,8 @@ public long putDocumentFromString(String xmlString, String uniqueID, String clus
             throws XmlServerException {
         try {
             InputSource source = new InputSource(new StringReader(xmlString));
-            Document document = DOCUMENT_BUILDER_FACTORY.newDocumentBuilder().parse(source);
+            //Document document = DOCUMENT_BUILDER_FACTORY.newDocumentBuilder().parse(source);
```
Comment (zbcui) [inline org.talend.mdm.core/src/com/amalto/core/storage/SystemStorageWrapper.java:343]: could remove

File: org.talend.mdm.core/src/com/amalto/core/storage/datasource/DataSourceFactory.java [hunk 5]
```diff
@@ -107,7 +110,13 @@ private static synchronized InputStream readDataSourcesConfiguration() {
     private static Map<String, DataSourceDefinition> readDocument(InputStream configurationAsStream) {
         Document document;
         try {
+            factory.setExpandEntityReferences(false);
+            factory.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
+            factory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, Boolean.TRUE);
+            factory.setFeature("http://xml.org/sax/features/external-general-entities", false);
+            factory.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
             DocumentBuilder documentBuilder = factory.newDocumentBuilder();
+//            DocumentBuilder documentBuilder = MDMXMLUtils.getDocumentBuilder().get();
```
Comment (zbcui) [inline org.talend.mdm.core/src/com/amalto/core/storage/datasource/DataSourceFactory.java:119]: could remove

**Review by zbcui — APPROVED** [review]

**Review by htyin — APPROVED** [review]

**Pull request merged**